### PR TITLE
Vladimir Shevtsov. 18-websockets

### PR DESCRIPTION
**Pull request by shevv920**

### CODE REVIEW

File: 18-websockets/classworks/Shevtsov_Vladimir/server/index.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+io.on('connection', (socket) => {
+  console.log(`New connection from: ${socket.id}`);
+
+  users.set(socket.id, 'anon');
+
+  socket.on('message', (data) => {
+    console.log(`Rcvd: message ${JSON.stringify(data)}`);
+    switch (data.type) {
+      case 'message':
```
Comment (Ivanko5417) [inline 18-websockets/classworks/Shevtsov_Vladimir/server/index.js]: it's better to use different events and add a handler for all of them:
```javascript 
socket.on('message', () => {
  //your code
});

socket.on('set_username', () => {
  //your code
});

socket.on('typing', () => {
  //your code
});
```

File: 18-websockets/classworks/Shevtsov_Vladimir/src/pages/Chat/Chat.jsx [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  const [username, setUsername] = useState('anon');
+  const [messages, setMessages] = useState([]);
+  const [typingUsers, setTypingUsers] = useState(new Set());
+
+  const addMessage = useCallback((msg) => {
+    setMessages([...messages, msg]);
+  }, [messages, setMessages]);
+
+  const addTyping = useCallback((data) => {
+    setTypingUsers((typing) => new Set([...typing, data.username]));
```
Comment (Ivanko5417) [inline 18-websockets/classworks/Shevtsov_Vladimir/src/pages/Chat/Chat.jsx]: looks like, if user doesn't send a message, the typing message won't disappear.
I'd suggest adding a timeout for 2-4 secs after getting the typing event for deleting user. If you get a message about typing again you should restart timeout.

File: 18-websockets/classworks/Shevtsov_Vladimir/src/pages/Chat/Chat.jsx [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  }, [inputValue, setInputValue, socket]);
+
+  const changeUsername = useCallback((event) => {
+    event.preventDefault();
+    socket.emit('message', { type: 'set_username', payload: username });
+  }, [username, socket]);
+
+  const handleInputChange = useCallback((event) => {
+    setInputValue(event.target.value);
+    socket.emit('message', { type: 'typing' });
```
Comment (Ivanko5417) [inline 18-websockets/classworks/Shevtsov_Vladimir/src/pages/Chat/Chat.jsx]: if the user is a super typer man, you will get a huge number of requests from him. A good practice is to use the `debounce` method for calling function no more than one time per 2-4 secs.
Unfortunately, javascript doesn't have this method. You can write it by yourself or use some library.

**Review by Ivanko5417 — CHANGES_REQUESTED** [review]
Good job!

File: 18-websockets/classworks/Shevtsov_Vladimir/src/pages/Chat/Chat.jsx (hunk 2 above)
Comment (shevv920) [reply 18-websockets/classworks/Shevtsov_Vladimir/src/pages/Chat/Chat.jsx]: fixed

File: 18-websockets/classworks/Shevtsov_Vladimir/src/pages/Chat/Chat.jsx (hunk 3 above)
Comment (shevv920) [reply 18-websockets/classworks/Shevtsov_Vladimir/src/pages/Chat/Chat.jsx]: fixed with lodash's debounce

**Review by Ivanko5417 — APPROVED** [review]
Beautiful!

**Pull request merged**